### PR TITLE
Overhaul vendoring/linking features

### PR DESCRIPTION
**Pull request by danielocfb**
Please see https://github.com/libbpf/libbpf-sys/pull/64 for a discussion and motivation and check individual commits for change descriptions.

### CODE REVIEW

**Comment by danielocfb** [comment]
Uhm, let me figure out the CI stuff.

**Comment by danielocfb** [comment]
@mmullin-halcyon I would think that in combination with [Alex' suggestion](https://github.com/libbpf/libbpf-sys/pull/64#issuecomment-1732173885) these changes should be able to support your use case, but I'd recommend you give them a try (and/or review).

**Comment by mmullin-halcyon** [comment]
I will look at this ASAP (probably tomorrow evening).

**Comment by danielocfb** [comment]
One more meta-comment: the current proposal enables dependent features as necessary. E.g., linking `libelf` statically enables static linking of `libbpf`. We could alternatively move to a model where we err out on combinations that make little sense/are risky/whatever. Doing so could be beneficial if we ever were to loosen some of those dependencies down the line to minimize chance of breakage. I am unsure how likely that really is or whether it would be cause for concern, and I generally like the declarative nature of "this feature depends on this other" in `Cargo.toml`. But we could change that if we feel strongly.

**Comment by mmullin-halcyon** [comment]
I am having a strange problem.

using
https://github.com/mmullin-halcyon/libbpf-rs/tree/origin/story/vendorize_dependencies (and changing a few flags)
and this feature-overhaul branch.  I am enabling both "static" and "vendored" (see bottom of comment)

I compile with the following command
```
RUSTFLAGS='-C target-feature=+crt-static' \
		cargo build --target x86_64-unknown-linux-gnu -r --features "vendored"
```
note: vendored for my binary is turning on static and vendored on libbpf-sys


My end binary is no longer static against libc.
```
ldd target/x86_64-unknown-linux-gnu/release/REDACTED
	linux-vdso.so.1 (0x00007ffe661bf000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f3fbee46000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f3fbed65000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f3fbe61e000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f3fbee81000)
```

From the log of libbpf-sys
```
Using feature vendored-libbpf=true
Using feature vendored-libelf=true
Using feature vendored-zlib=true
Using feature static-libbpf=true
Using feature static-libelf=true
Using feature static-zlib=true
```

I can't see anywhere in the code which should cause this problem.  Please give me some more time before submitting to sus-out what is going on.

File: build.rs [hunk 1]
```diff
@@ -253,9 +252,6 @@ fn make_elfutils(compiler: &cc::Tool, src_dir: &path::PathBuf, out_dir: &path::P
             }
         })
         .collect();
-
-    #[cfg(target_arch = "aarch64")]
```
Comment (mmullin-halcyon) [inline build.rs:257]: Put the aarch64 kludge back in please

**Review by mmullin-halcyon — CHANGES_REQUESTED** [review]
I did not test on AARCH64, but I suspect that the libelf static compile kludge on that platform is still needed.

File: build.rs (hunk 1 above)
Comment (danielocfb) [reply build.rs:257]: Ah okay. I should not have reverted that change then. Ignoring it now.

**Comment by mmullin-halcyon** [comment]
LGTM

File: src/lib.rs [hunk 2]
```diff
@@ -6,7 +6,7 @@
 
 include!("bindings.rs");
 
-#[cfg(feature = "static")]
+#[cfg(feature = "vendor-libbpf")]
```
Comment (alexforster) [inline src/lib.rs:9]: Looks like there's some mismatch around feature naming: `vendor-blah` vs `vendored-blah`.

File: .github/workflows/ci.yml
Comment (alexforster) [inline .github/workflows/ci.yml:1]: I'm not fluent in Github CI – does the "matrix" now test all the various feature combinations?

**Comment by alexforster** [comment]
This is awesome! Left some questions/comments.

> We could alternatively move to a model where we err out on combinations that make little sense/are risky/whatever.
> I generally like the declarative nature of "this feature depends on this other" in `Cargo.toml`.

I like it too, I'm fine with it the way it is.

File: src/lib.rs (hunk 2 above)
Comment (danielocfb) [reply src/lib.rs]: Oh yes, that's a good catch. Thanks, fixed.

File: .github/workflows/ci.yml
Comment (danielocfb) [reply .github/workflows/ci.yml:1]: It's only an adjustment of the combinations we already build. That's what I tried to convey with

> I tested everything on a binary depending on libbpf-sys with various
> features enabled and spot checked the expected dynamic library
> dependencies. We could enshrine that as a CI step, but given that this
> logic is expected to change infrequently I didn't go down that road.

Do let me know if you want to have all possible combinations tested in CI and to what degree.

Comment (alexforster) [reply .github/workflows/ci.yml:1]: It's okay, I was just trying to make sure I understood what was changing.

**Comment by alexforster** [comment]
I'm happy if y'all are happy. Ready to merge?

**Comment by danielocfb** [comment]
Hi @alexforster, would you be able to cut a release containing the changes and the [`libbpf` bump](https://github.com/libbpf/libbpf-sys/pull/72)?

**Pull request merged**